### ISSUE TITLE
DRBD net protocol may be selectable ?

### ISSUE THREAD

**Issue by Ganeti-Issues-Migrator** [ganeti/ganeti#242]
Originally reported of Google Code with ID 186.

```
I would like to know if there is a reason why the DRBD_NET_PROTOCOL is "hard-coded" "C" in constants.py ?

Is there any reason like "we must have a zero data loss solution" ?

I want to use protocol A, for performance reasons. (Gigabit network is not fast enough for me), so I just modified the constants.py.

But I just want to do something smarter than this.

So my question is, should I write a patch to support setting the net_protocol or will it be rejected cause ganeti don't want anything other than C protocol ?

Thanks,

```
Originally added on 2011-08-22 09:41:40 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
The answer for the hardcoding of C is that we internally only use C :)

If you want to use protocol A, feel free. A patch just adding support for this is not good because we want to add generic support for DRBD parameters, as opposed to just toggling the protocol.
```
Originally added on 2011-08-22 10:00:35 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
But if you were to implement it with disk parameters, like network parameters but for drbd params (protocol, speed, etc) it would definitely be accepted.
```
Originally added on 2011-08-23 09:14:53 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Yeah ok, I'll work on that this week end.
```
Originally added on 2011-08-23 12:22:40 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
During migration, DRBD needs to be put in multimaster in order to start both qemu-kvm instances at the same time.

So, DRBD *NEEDS* to be protocol C at this stage.

But for now, if constants.py DRBD_NET_PROTOCOL = "A", then the migration will fail, cause drbd will not accept dual master in protocol other than C.

That's why I've written a very little patch against 2.4.4 which handle this case.


What do you think of it?
```
Originally added on 2011-08-25 11:48:10 +0000 UTC.

Attachments: 
[ganeti.force_protocol_c_for_migration.diff](https://storage.googleapis.com/google-code-attachments/ganeti/issue-186/comment-4/ganeti.force_protocol_c_for_migration.diff)

**Comment by Ganeti-Issues-Migrator** [comment]
```
Will we also need to bring the protocol back to A or B after the migration is finished?

Thanks,

Guido
```
Originally added on 2012-07-10 00:14:20 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Guido: yes, we do; otherwise it will be A/B only for a short while, if the cluster is rebalanced :)
```
Originally added on 2012-07-10 06:40:28 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
We use protocol A in production / "normal" mode, but C just for migration. It works really well, as performance impact of protocol C during the migration is not a big deal. Usually when you do migration, you are aware of the performance impact.

The performance impact is minimal, and anyway without the patch, you're just stuck with C so.. ;)

As of now, we are very happy of the patch posted.

Anyway, how can you do sync with protocol other than C ?
```
Originally added on 2012-07-10 06:57:31 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
How I did not fully understood the question, so yes, need to bring back to A after migration ! :-)
```
Originally added on 2012-07-10 06:58:49 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
True. If needed I can help test this, as I use the patch posted above in production.

Thank you for your hard work :-)
```
Originally added on 2012-07-10 06:59:51 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
-- Empty comment --
```
Originally added on 2012-07-19 14:34:57 +0000 UTC.

Added Labels: Type-Enhancement Component-storage

**Comment by Ganeti-Issues-Migrator** [comment]
```
Hi Laurent,

Could you please sign the code.google.com CLA?
Then I believe we can import this patch, and any further one to set the sync protocol via parameters.

Thanks,

Guido

```
Originally added on 2013-02-13 14:38:49 +0000 UTC.

Changed State: Accepted

**Comment by Ganeti-Issues-Migrator** [comment]
```
-- Empty comment --
```
Originally added on 2013-05-08 08:30:34 +0000 UTC.

Added to Milestone: Release2.9

**Comment by Ganeti-Issues-Migrator** [comment]
```
Oh sorry, I missed the request for CLA. I'm filling it right now.
I still use the patch in production since I've done it :)
```
Originally added on 2013-05-08 11:48:39 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Fixed in Ganeti 2.9 with:

commit 76e110235a18283c534d5939ebe06ec0bd911fc5
Author: Thomas Thrainer <thomasth@google.com>
Date:   Tue Jul 2 15:47:26 2013 +0200

    Document the 'protocol' disk parameter
    
    Document the new disk parameter 'protocol' which is used to set the DRBD
    replication protocol.
    
    Signed-off-by: Thomas Thrainer <thomasth@google.com>
    Reviewed-by: Klaus Aehlig <aehlig@google.com>

commit 65fc238843b18ddd460c8f811ab45aff4eddbd3e
Author: Thomas Thrainer <thomasth@google.com>
Date:   Tue Jul 2 13:56:03 2013 +0200

    Make DRBD replication protocol configurable
    
    Add a new disk parameter which allows to choose the DRBD replication
    protocol. The protocol is fixed to "C" during dual primary mode though.
    
    Signed-off-by: Thomas Thrainer <thomasth@google.com>
    Reviewed-by: Klaus Aehlig <aehlig@google.com>

```
Originally added on 2013-07-03 06:53:49 +0000 UTC.

Changed State: Fixed

**Comment by Ganeti-Issues-Migrator** [comment]
```
Ganeti 2.9.0 has been released
```
Originally added on 2013-11-05 13:56:36 +0000 UTC.

Changed State: Released
